### PR TITLE
[fix]: [CDS-50144]: Add validation for pipeline runtime inputs in trigger

### PR DESCRIPTION
**Pull request by rathodmeetsatish**
## Describe your changes
Add validation for pipeline runtime inputs in trigger

## Checklist
- [ ] I've documented the changes in the PR description.
- [ ] I've tested this change either in PR or local environment.
- [ ] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

## Comment Triggers
<details>
  <summary>Build triggers</summary>

- Feature build: `trigger feature-build`
- Immutable delegate `trigger publish-delegate`
</details>

<details>
  <summary>PR Check triggers</summary>

You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ti0, ti1`

- Compile: `trigger compile`
- CodeformatCheckstyle: `trigger checkstylecodeformat`
    - CodeFormat: `trigger codeformat`
    - Checkstyle: `trigger checkstyle`
- MessageMetadata: `trigger messagecheck`
- File-Permission-Check: `trigger checkpermission`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- PMD: `trigger pmd`
- Copyright Check: `trigger copyrightcheck`
- Feature Name Check: `trigger featurenamecheck`
- TI-bootstrap: `trigger ti0`
- TI-bootstrap1: `trigger ti1`
- TI-bootstrap2: `trigger ti2`
- TI-bootstrap3: `trigger ti3`
- TI-bootstrap4: `trigger ti4`
- FunctionalTest1: `trigger ft1`
- FunctionalTest2: `trigger ft2`
- CodeBaseHash: `trigger codebasehash`
- CodeFormatCheckstyle: `trigger checkstylecodeformat`
- SonarScan: `trigger ss`
- GitLeaks: `trigger gitleaks`
- Trigger all Checks: `trigger smartchecks`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions

## [Contributor license agreement](https://github.com/harness/harness-core/blob/develop/CONTRIBUTOR_LICENSE_AGREEMENT.md)

- - -
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/43374)

### CODE REVIEW

File: pipeline-service/modules/ng-triggers/src/main/java/io/harness/ngtriggers/service/impl/NGTriggerServiceImpl.java [hunk 1]
```diff
@@ -833,6 +833,15 @@ public Map<FQN, String> getInvalidFQNsInTrigger(String templateYaml, String trig
       return errorMap;
     }
 
+    templateConfig.getFqnToValueMap().keySet().forEach(key -> {
+      Object templateValue = templateConfig.getFqnToValueMap().get(key);
+      if (templateValue.equals(TextNode.valueOf("<+input>"))) {
+        if (!triggerFQNs.contains(key)) {
```
Comment (vcalasansh) [inline pipeline-service/modules/ng-triggers/src/main/java/io/harness/ngtriggers/service/impl/NGTriggerServiceImpl.java:839]: Can't we just move this logic inside the `else` block starting on line 865?

Also, does this code block currently starting on line 865 have any purpose at all? Can we just replace it?
```
 Map<FQN, Object> subMap = YamlSubMapExtractor.getFQNToObjectSubMap(triggerConfig.getFqnToValueMap(), key);
        subMap.keySet().forEach(triggerFQNs::remove);
```

File: pipeline-service/modules/ng-triggers/src/main/java/io/harness/ngtriggers/service/impl/NGTriggerServiceImpl.java [hunk 2]
```diff
@@ -833,6 +833,15 @@ public Map<FQN, String> getInvalidFQNsInTrigger(String templateYaml, String trig
...
+          errorMap.put(key, "Trigger does not contain pipeline runtime input" + key.getExpressionFqn());
```
Comment (vcalasansh) [inline pipeline-service/modules/ng-triggers/src/main/java/io/harness/ngtriggers/service/impl/NGTriggerServiceImpl.java:840]: Please add a white space at the end of the string here

**Review by rathodmeetsatish — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/43374)* status: 1 file(s) to review. 1. unresolved discussions. 0 of 2 approvals obtained. 

___
*[`pipeline-service/modules/ng-triggers/src/main/java/io/harness/ngtriggers/service/impl/NGTriggerServiceImpl.java` line 839 at r1](https://reviewable.io/reviews/harness/harness-core/43374#-NNCklu69hS2Kfig2iit-r1-839:-NNCxTn-4U62iLt35NTW:bml1tw) ([raw file](https://github.com/harness/harness-core/blob/8ace16d3bcc5af7596f58f1c3a02e32386072311/pipeline-service/modules/ng-triggers/src/main/java/io/harness/ngtriggers/service/impl/NGTriggerServiceImpl.java#L839)):*
<details><summary><i>Previously, vcalasansh (Vinícius Calasans) wrote…</i></summary><blockquote>

Can't we just move this logic inside the `else` block starting on line 865?

Also, does this code block currently starting on line 865 have any purpose at all? Can we just replace it?
```
 Map<FQN, Object> subMap = YamlSubMapExtractor.getFQNToObjectSubMap(triggerConfig.getFqnToValueMap(), key);
        subMap.keySet().forEach(triggerFQNs::remove);
```
</blockquote></details>

as mentioned in comment on 845 line ,the 846 to 869 block is it to check whether the trigger have some extra field that pipeline doesnt have and the block that i have add tests that if runtime input are provided or not

**Comment by rathodmeetsatish** [comment]
trigger messagecheck

**Review by rathodmeetsatish — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/43374)* status: 6 file(s) to review. 2. unresolved discussions. 0 of 2 approvals obtained. 

___
*[`pipeline-service/modules/ng-triggers/src/main/java/io/harness/ngtriggers/service/impl/NGTriggerServiceImpl.java` line 840 at r1](https://reviewable.io/reviews/harness/harness-core/43374#-NNCklu69hS2Kfig2iit-r1-840:-NNbuRtf3I3nfZzsHQFE:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/8ace16d3bcc5af7596f58f1c3a02e32386072311/pipeline-service/modules/ng-triggers/src/main/java/io/harness/ngtriggers/service/impl/NGTriggerServiceImpl.java#L840)):*
<details><summary><i>Previously, vcalasansh (Vinícius Calasans) wrote…</i></summary><blockquote>

Please add a white space at the end of the string here
</blockquote></details>

Done.

File: 956-feature-flag-beans/src/main/java/io/harness/beans/FeatureName.java [hunk 3]
```diff
@@ -617,7 +617,8 @@ public enum FeatureName {
       HarnessTeam.CDP),
   CI_PIPELINE_VARIABLES_IN_STEPS("For enabling pipeline variables as env variables in CI steps", HarnessTeam.CI),
   CDS_ENTITY_REFRESH_DO_NOT_QUOTE_STRINGS(
-      "Do not add quotes to strings when a user reconciles a template, pipeline", HarnessTeam.CDC, Scope.GLOBAL);
+      "Do not add quotes to strings when a user reconciles a template, pipeline", HarnessTeam.CDC, Scope.GLOBAL),
+  SPG_VALIDATE_PIPELINE_RUNTIME_INPUT_FOR_TRIGGER("Validate the pipeline runtime input for triggers", HarnessTeam.SPG);
```
Comment (vcalasansh) [inline 956-feature-flag-beans/src/main/java/io/harness/beans/FeatureName.java:621]: Manager (the source of FFs) and pipeline-svc don't have the same release schedule - You will need to first merge a PR adding the FF and wait for it to be released before merging the other changes.

File: pipeline-service/modules/ng-triggers/src/main/java/io/harness/ngtriggers/service/impl/NGTriggerServiceImpl.java [hunk 4]
```diff
@@ -871,6 +866,10 @@ public Map<FQN, String> getInvalidFQNsInTrigger(String templateYaml, String trig
       } else {
         Map<FQN, Object> subMap = YamlSubMapExtractor.getFQNToObjectSubMap(triggerConfig.getFqnToValueMap(), key);
         subMap.keySet().forEach(triggerFQNs::remove);
+        if (pmsFeatureFlagService.isEnabled(
```
Comment (vcalasansh) [inline pipeline-service/modules/ng-triggers/src/main/java/io/harness/ngtriggers/service/impl/NGTriggerServiceImpl.java:869]: Could you please make sure that the response to this call to `isEnabled` is cached?

File: pipeline-service/modules/ng-triggers/src/main/java/io/harness/ngtriggers/service/impl/NGTriggerServiceImpl.java [hunk 5]
```diff
@@ -778,7 +771,8 @@ public Map<String, Map<String, String>> validatePipelineRef(TriggerDetails trigg
         String templateYaml = createRuntimeInputForm(pipelineYaml);
         String triggerYaml = triggerDetails.getNgTriggerEntity().getYaml();
         String triggerPipelineYml = getPipelineComponent(triggerYaml);
-        Map<FQN, String> invalidFQNs = getInvalidFQNsInTrigger(templateYaml, triggerPipelineYml);
+        String accountIdentifier = triggerDetails.getNgTriggerEntity().getAccountId();
+        Map<FQN, String> invalidFQNs = getInvalidFQNsInTrigger(templateYaml, triggerPipelineYml, accountIdentifier);
```
Comment (vcalasansh) [inline pipeline-service/modules/ng-triggers/src/main/java/io/harness/ngtriggers/service/impl/NGTriggerServiceImpl.java:775]: Were we not also removing the call to `validatePipelineRef` from update/create endpoints when `ignoreError=true`?

**Review by rathodmeetsatish — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/43374)* status: 6 file(s) to review. 4. unresolved discussions. 0 of 2 approvals obtained. 

___
*[`pipeline-service/modules/ng-triggers/src/main/java/io/harness/ngtriggers/service/impl/NGTriggerServiceImpl.java` line 775 at r3](https://reviewable.io/reviews/harness/harness-core/43374#-NNCklu69hS2Kfig2iit-r3-775:-NNdwo8L1Ck7dQTY86VG:b5yds17) ([raw file](https://github.com/harness/harness-core/blob/5f18bf775d6efac58c9450f4c34018ae75464930/pipeline-service/modules/ng-triggers/src/main/java/io/harness/ngtriggers/service/impl/NGTriggerServiceImpl.java#L775)):*
<details><summary><i>Previously, vcalasansh (Vinícius Calasans) wrote…</i></summary><blockquote>

Were we not also removing the call to `validatePipelineRef` from update/create endpoints when `ignoreError=true`?
</blockquote></details>

yes we have done that 
___
*[`pipeline-service/modules/ng-triggers/src/main/java/io/harness/ngtriggers/service/impl/NGTriggerServiceImpl.java` line 869 at r3](https://reviewable.io/reviews/harness/harness-core/43374#-NNCklu69hS2Kfig2iit-r3-869:-NNdwvYJEeNIj0KlK8cg:b-skotjw) ([raw file](https://github.com/harness/harness-core/blob/5f18bf775d6efac58c9450f4c34018ae75464930/pipeline-service/modules/ng-triggers/src/main/java/io/harness/ngtriggers/service/impl/NGTriggerServiceImpl.java#L869)):*
<details><summary><i>Previously, vcalasansh (Vinícius Calasans) wrote…</i></summary><blockquote>

Could you please make sure that the response to this call to `isEnabled` is cached? 
</blockquote></details>

i checked isEnabled methods implementation its already cached

**Review by vcalasansh — APPROVED** [review]
> Reviewed 4 of 6 files at r2, 1 of 1 files at r3, 1 of 1 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/43374)* status: 6 file(s) to review. 1 of 2 approvals obtained.

**Comment by rathodmeetsatish** [comment]
trigger ti0

**Comment by rathodmeetsatish** [comment]
trigger ti2 trigger ti3 trigger ti4

**Review by vcalasansh — APPROVED** [review]
> Reviewed 1 of 1 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/43374)* status: 6 file(s) to review. 1 of 2 approvals obtained.

**Comment by rathodmeetsatish** [comment]
trigger ti1 trigger ti2

**Comment by rathodmeetsatish** [comment]
trigger ti1

**Comment by vcalasansh** [comment]
trigger tin1

**Review by sribalij — DISMISSED** [review]
> Reviewed 3 of 6 files at r2, 1 of 1 files at r5, 2 of 2 files at r6, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/43374)* status: 2 file(s) to review. 1 of 2 approvals obtained.

**Review by vcalasansh — DISMISSED** [review]
> Reviewed 2 of 2 files at r6, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/43374)* status: 6 file(s) to review. 1 of 2 approvals obtained.

**Comment by rathodmeetsatish** [comment]
trigger ti1

**Review by RajBaviskar — DISMISSED** [review]
:lgtm:

> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/43374#-:-NNwiy8R6trofLI-h-fj:bnfp4nl)* status: 6 file(s) to review.

**Review by vcalasansh — APPROVED** [review]
> Reviewed 1 of 1 files at r7, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/43374)* status: 6 file(s) to review. 1 of 2 approvals obtained.

**Review by rijajoo — APPROVED** [review]
> Reviewed 3 of 6 files at r2, 2 of 2 files at r6, 1 of 1 files at r7, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/43374)* status: :shipit: complete! ✓

**Pull request merged**